compiler/rustc_target/src/asm/sparc.rs

```diff
@@ -0,0 +1,135 @@
+use std::fmt;
+
+use rustc_data_structures::fx::FxIndexSet;
+use rustc_span::{Symbol, sym};
+
+use super::{InlineAsmArch, InlineAsmType, ModifierInfo};
+use crate::spec::{RelocModel, Target};
+
+def_reg_class! {
+    Sparc SparcInlineAsmRegClass {
+        reg,
+        yreg,
+    }
+}
+
+impl SparcInlineAsmRegClass {
+    pub fn valid_modifiers(self, _arch: super::InlineAsmArch) -> &'static [char] {
+        &[]
+    }
+
+    pub fn suggest_class(self, _arch: InlineAsmArch, _ty: InlineAsmType) -> Option<Self> {
+        None
+    }
+
+    pub fn suggest_modifier(
+        self,
+        _arch: InlineAsmArch,
+        _ty: InlineAsmType,
+    ) -> Option<ModifierInfo> {
+        None
+    }
+
+    pub fn default_modifier(self, _arch: InlineAsmArch) -> Option<ModifierInfo> {
+        None
+    }
+
+    pub fn supported_types(
+        self,
+        arch: InlineAsmArch,
+    ) -> &'static [(InlineAsmType, Option<Symbol>)] {
+        match self {
+            Self::reg => {
+                // FIXME: i64 is ok for g*/o* registers on SPARC-V8+ ("h" constraint in GCC)
+                if arch == InlineAsmArch::Sparc {
+                    types! { _: I8, I16, I32; }
+                } else {
+                    types! { _: I8, I16, I32, I64; }
+                }
+            }
+            Self::yreg => &[],
+        }
+    }
+}
+
+pub(crate) fn is_v7_or_v8(arch: InlineAsmArch, target_features: &FxIndexSet<Symbol>) -> bool {
+    // FIXME: also check v8plus in LLVM 20+:
+    // https://github.com/llvm/llvm-project/commit/aca971d336d9c7650120fc0fd6dfe58866408216
+    arch == InlineAsmArch::Sparc && !target_features.contains(&sym::v9)
+}
+
+fn reserved_g5(
+    arch: InlineAsmArch,
+    _reloc_model: RelocModel,
+    target_features: &FxIndexSet<Symbol>,
+    _target: &Target,
+    _is_clobber: bool,
+) -> Result<(), &'static str> {
+    if is_v7_or_v8(arch, target_features) {
```

Comment:
The condition used in LLVM is `Subtarget.is64Bit()`, not whether `v8plus` is supported. So this should check for `InlineAsmArch::Sparc64` instead.

Reply:
Hmm. Section 2.1.5 "Function Registers with Unassigned Roles" of the V8+ Technical Specification says "%g5; no longer reserved for system software" [1], and LLVM does not seem to match it.

I changed the condition to Sparc64 vs Sparc and left a FIXME comment mentioning the mismatch between LLVM and the specification.

[1]: https://temlib.org/pub/SparcStation/Standards/V8plus.pdf

Reply:
(The above change removes the need for the addition of the target feature in this PR. It is still needed for other purposes but will be handled in another PR: https://github.com/rust-lang/rust/pull/132552)